File: day69/buildTree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

# class Solution:
#     def buildTree(self, preorder: List[int], inorder: List[int]) -> TreeNode:
#         if not preorder: return None

#         root = TreeNode(preorder[0])
#         index = inorder.index(root.val)
#         root.left = self.buildTree(preorder[1: index + 1], inorder[:index])
#         root.right = self.buildTree(preorder[index + 1:], inorder[index + 1:])
#         return root


class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> TreeNode:
        if not preorder: return None

        def recursion(l, r, f, e):
            if l == r or f == e: return None

            root = TreeNode(preorder[l])
            index = inorder.index(root.val) - f
            root.left = recursion(l + 1, l + index + 1, f, f + index)
            root.right = recursion(l + index + 1, r, f + index + 1, e)
            return root
        
        return recursion(0, len(preorder), 0, len(inorder))

File: day69/buildTree.java
/*
面试题07. 重建二叉树
输入某二叉树的前序遍历和中序遍历的结果，请重建该二叉树。假设输入的前序遍历和中序遍历的结果中都不含重复的数字。

链接: https://leetcode-cn.com/problems/zhong-jian-er-cha-shu-lcof/
*/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {

    private int[] preorder;
    private Map<Integer, Integer> map = new HashMap<>();

    public TreeNode buildTree(int[] preorder, int[] inorder) {
        if (preorder == null || preorder.length == 0) return null;
        this.preorder = preorder;
        for (int i = 0; i < inorder.length; i++) {
            this.map.put(inorder[i], i);
        }

        return recursion(0, preorder.length, 0, inorder.length);

    }

    public TreeNode recursion(int l ,int r, int f, int e) {
        if (l == r || f == e) return null;

        TreeNode root = new TreeNode(this.preorder[l]);
        int i = this.map.get(this.preorder[l]) - f;
        root.left = recursion(l + 1, l + 1 + i, f, f + i);
        root.right = recursion(l + 1 + i, r, f + 1 + i, e);

        return root;
    }
}

